### ISSUE TITLE
[Improvement] Improve the current SQL of JDBCBackend

### ISSUE THREAD

**Issue by jerryshao** [datastrato/gravitino#4024]
### What would you like to be improved?

In the current implementation of xxxMetaMapper and xxxMetaService, we seldom use JOIN and transactions, for example:

<img width="790" alt="image" src="https://github.com/datastrato/gravitino/assets/850797/4b6fe9c0-f30e-4f21-bf3c-e6a964066ede">

The list catalog operation will issue two SQL queries, the first one is to get a metalake id, then using this id to get the catalog list.

Instead of issuing two queries, we can use JOIN to join catalog and metalake table to get catalog list by one query, this will save the IO time and avoid inconsistent problem.

Also for `updateCatalog`

<img width="1005" alt="image" src="https://github.com/datastrato/gravitino/assets/850797/568e468e-9c25-4ec9-b4ae-e4112bc3123e">

We have several steps:

1. get metalake id.
2. get catalog PO.
3. update the catalog object.
4. update the catalog in RDBMS.

These 4 steps are not in the transaction, which will potentially meet the inconsistent problem, and we highly leverage the lock to avoid the concurrent problem. A better solution is to put these 4 steps into a transaction.

### How should we improve?

So basically, we can improve the current SQL to:

1. reduce the queries and increase the performance.
2. Put multiple step actions into one transaction to leverage DB's transaction to keep consistent.

**Comment by jerryshao** [comment]
CC @xloya @yuqi1129 , please take a look when you have time.

**Comment by jerryshao** [comment]
It's a big work to change everything, we can separate it into small tasks to iterate one by one. Currently, there's no functionality issue here since we have locks to guarantee the concurrency.

**Comment by xloya** [comment]
1. I think there is no problem with using JOIN to reduce the number of queries.
2. As for whether to put all operations in the same transaction, I may have some doubts, because this may cause a database transaction to be too large and cause deadlock and other problems.
3. I think we can consider whether it is enough to ensure transactionality with upper-level locks (later, while supporting horizontal expansion of the server, ensuring transactionality), or whether a large database transaction is really needed.

**Comment by jerryshao** [comment]
I have no clear answer for now. I think we can do some tests to see if transaction is OK or not. But for join I think we can do some changes somehow. For example, like delete operation, we can use join to do it and check the return value, you can check my code #4019 , I use join to handle some cases. @xloya

**Comment by yuqi1129** [comment]
> Hi [<img alt="" width="16" height="16" src="https://avatars.githubusercontent.com/u/850797?u=fabb5216d8d77522b02ff6ac72ede0fdbd95b7cf&amp;v=4&amp;size=80">@jerryshao](https://github.com/jerryshao?rgh-link-date=2025-02-14T12%3A36%3A03.000Z) Can I work on: using JOIN to reduce the number of queries?

Sure, please go ahead if you are interested in this issue.
